### PR TITLE
first commit for Force Control documentation

### PR DESCRIPTION
**Pull request by davidetome**
This PR aims to start a discussion on what material have to be migrated/updated.

cc @pattacini @traversaro

### CODE REVIEW

**Review by pattacini — CHANGES_REQUESTED** [review]
## MoM of the meeting held between @davidetome @traversaro and @pattacini

We decided to proceed in 2 steps.

### Step 1
- Migrate the wiki page http://wiki.icub.org/wiki/Force_Control first.
- Remove the table as it is deprecated now.
- Keep the description of the control modes as they're quite useful although we will very likely move them to a different page in the near future.
- Replace all the old machinery with the latest one.

### Step 2
- Deal with the device-based WBD.

**Comment by davidetome** [comment]
As [agreed](https://github.com/icub-tech-iit/documentation/pull/201#pullrequestreview-905921252) , I made some mods

File: docs/icub_force_control/icub-force-control.md [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+
+The key concept to comprehend how force control works and how to use it is the concept of control mode. The control mode represents the current control algorithm that is running on the firmware of the control boards to control a specific joint. For example, the position control modes implements a PID control that tracks the commanded trajectories, while the impedance control realizes a compliant position control by computing the reference torque that an inner torque control loop has to track, given an equilibrium position and the stiffness of a simulated spring. The control mode of a joint can be changed online, during the execution of your application, using the apposite yarp interfaces. In this way you can assign different control modes to different joints in order to obtain the desired behaviour (e.g. you can set some joints in position control mode to obtain a 'stiff' behaviour and other joint in impedance control mode to obtain a compliant behaviour).
+
+Five different control modes are currently implemented in the firmware of the control boards:
+
+#### Position control mode <font color="#95ddba">███</font></li></ul>
```
Comment (pattacini) [inline docs/icub_force_control/icub-force-control.md:163]:
```suggestion
#### Position control mode
```

File: docs/icub_force_control/icub-force-control.md [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+
+(Typical input parameters: desired position, trajectory velocity)
+Position control is the standard control mode. In this control mode, the motors PWM is computed using a PID controller the receives in input the desired joint position and the current measurement from the joint encoders:
+
+![pid-pos](./img/Pid_pos.jpg)
+
+Note that when you command a new joint position, you are not instantaneously assigning the reference qd in the above formula. Instead, a mimum jerk trajectory generator takes in input your commanded position and the desired velocity, and produces a smooth movement creating a sequence of position references qd tracked by the PID controller.
+
+#### Velocity control mode <font color="#96beff">███</font></li></ul>
```
Comment (pattacini) [inline docs/icub_force_control/icub-force-control.md:172]:
```suggestion
#### Velocity control mode
```

File: docs/icub_force_control/icub-force-control.md [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+(Typical input parameters: desired velocity, acceleration)
+Velocity control mode allows you to control the robot by assigning a desired volocity/acceleration to a joint. The control law is the same of position control, but in this case qd is not directly controlled by the user, but it is obtained from the integration of the commanded user velocity. Also in this case a minimum jerk profile generator is used.
+
+#### Torque control mode <font color="#ff6464">███</font></li></ul>
```
Comment (pattacini) [inline docs/icub_force_control/icub-force-control.md:176]:
```suggestion
#### Torque control mode
```

File: docs/icub_force_control/icub-force-control.md [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+(Typical input parameters: reference torque)
+Torque control mode allows you to directly control the robot joints torque:
+Pid trq.jpg
+In this case the motors PWM is computed using a PID controller the receives in input the desired joint torque and the current measured joint torque. Additionally, a PWM offset can be added to the output of the control algorithm. If both the commanded reference torque and the PWM offset is set to zero, the robot joint will be free to be moved in the space (eventually it will move down as an effect of the gravity acting on that joint).
+
+![pid-pos](./img/Pid_trq.jpg)
+
+#### Openloop control mode <font color="#000000">░░░</font></li></ul>
```
Comment (pattacini) [inline docs/icub_force_control/icub-force-control.md:184]:
```suggestion
#### Openloop control mode
```

File: docs/icub_force_control/icub-force-control.md [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+(Typical input parameters: motor PWM)
+Openloop control mode allows you to directly control the joint motor, assigning directly the PWM (bypassing the PID controller)
+
+#### Impedance Position control mode <font color="#dcbedc">███</font></li></ul>
```
Comment (pattacini) [inline docs/icub_force_control/icub-force-control.md:188]:
```suggestion
#### Impedance Position control mode
```

File: docs/icub_force_control/icub-force-control.md [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+(Typical input parameters: desired position, trajectory velocity + desired joint stiffness and damping)
+Impedance control mode allows you to control the joint position and its compliance. In particular, you can control the equilibrium position of a virtual spring (using the standard yarp::dev::IPositionControl interface) and its stiffness/damping (using the yarp::dev::IImpedanceControl interface).
+The control is implemented in the DSP firmware as follows:
+Pid imp.jpg
+Firstly, a reference torque is computed, accordingly to the input position and the commanded stiffness/damping parameters (Hooke's law). Secondly, the reference torque is tracked by a PID algorithm (same gains used by the torque control mode). By tuning the stiffness parameters, you can thus make the robot joint feeling like a hard or soft spring, while maintaining control on the desired joint position (note that the same mimum-jerk trajectory generator used by the position control is also used when position impedance control is running).
+
+![pid-pos](./img/Pid_imp.jpg)
+
+#### Impedance Velocity control mode <font color="#dcbedc">███</font></li></ul>
```
Comment (pattacini) [inline docs/icub_force_control/icub-force-control.md:197]:
```suggestion
#### Impedance Velocity control mode
```

File: docs/icub_force_control/icub-force-control.md [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+(Typical input parameters: desired velocity, acceleration + desired joint stiffness and damping)
+The impedance velocity control mode is the corresponding impedance mode using velocity control. The control law is the same of the impedance position control, but in this case qd is not directly controlled by the user, but it is obtained from the integration of the commanded user velocity (also in this case minimum jerk profile generator is used).
+
+#### Idle <font color="#f9ec8d">███</font></li></ul>
```
Comment (pattacini) [inline docs/icub_force_control/icub-force-control.md:201]:
```suggestion
#### Idle
```

File: docs/icub_force_control/icub-force-control.md [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+⚠️ Be sure to move the part applying the force by hand respecting the following :
+
+|Part|Point where to apply the force|
+|---|---|
+| Arms  | Forearm/wrist| 
+|  Legs | Ankle|
+
+
+### Ports and connections
+wholeBodyDynamics and `yarprobotinterface`` communicate through yarp ports:
```
Comment (pattacini) [inline docs/icub_force_control/icub-force-control.md:140]:
```suggestion
wholeBodyDynamics and `yarprobotinterface` communicate through yarp ports:
```

File: docs/icub_force_control/icub-force-control.md [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+
+iCub ports are:
+
+/icub/inertial provides 3DOF orientation tracker measurements
```
Comment (pattacini) [inline docs/icub_force_control/icub-force-control.md:144]:
```suggestion
`/icub/inertial` provides 3DOF orientation tracker measurements
```
I would enclose all port names within grave accents in order to improve the formatting.

**Review by pattacini — COMMENTED** [review]
Just a few initial comments @davidetome.

I've proposed commits to fix some section titles where colors show up. The index gets messed up indeed:

image

I'd highlight the port names using this `markdown formatting`.

Did you also make sure to delete images that are no longer used?

**Review by pattacini — CHANGES_REQUESTED** [review]
Hi @davidetome 

I can see only 3 images of the `yarpmanager` rendered: one seems to be missing.

Also, all the images should be rendered approximately with the same size but this is not the case.

**Comment by davidetome** [comment]
should be fixed 👍🏻

**Review by pattacini — APPROVED** [review]
Nicely done @davidetome 👍🏻 
I deem we're done with step 1.
What do you think @traversaro ?

**Comment by davidetome** [comment]
Ciao @pattacini , about Step 2,  I already ported [this page](https://github.com/robotology/whole-body-estimators/blob/master/doc/howto/force_control_on_icub.md) on the top of the document, do you think something else is missing? @traversaro what do you think?

**Comment by pattacini** [comment]
>on the top of the document

On the top is not the correct place. We said that the classical module-based WBD is still the official one, hence it has to receive the focus.

Let's put it at the bottom then.

**Comment by davidetome** [comment]
> > on the top of the document
> 
> The top is not the correct place. We said that the classical module-based WBD is still the official one, hence it has to receive the focus.
> 
> Let's put the service-based WBD at the bottom then.

I thought the same, I agree 👍🏻

**Comment by traversaro** [comment]
I agree!

**Comment by davidetome** [comment]
@pattacini @traversaro , I did the mods and added some missing parts..

**Comment by pattacini** [comment]
The failure in the CI is not related to the content of this PR.
cc @davidetome

**Comment by pattacini** [comment]
> The failure in the CI is not related to the content of this PR. cc @davidetome

The GA framework is operational again. I'll merge the PR then.

**Pull request merged**